src/FirebaseAuth.js

```diff
@@ -246,27 +245,41 @@
      * rejected if the client is unauthenticated and rejectIfAuthDataIsNull is true.
      */
     _routerMethodOnAuthPromise: function(rejectIfAuthDataIsNull) {
-      var ref = this._ref;
+      var ref = this._ref, utils = this._utils;
+      // wait for the initial auth state to resolve; on page load we have to request auth state
+      // asynchronously so we don't want to resolve router methods or flash the wrong state
+      return this._initialAuthResolver.then(function() {
+        // auth state may change in the future so rather than depend on the initially resolved state
+        // we also check the auth data (synchronously) if a new promise is requested, ensuring we resolve
+        // to the current auth state and not a stale/initial state (see https://github.com/firebase/angularfire/issues/590)
+        var authData = ref.getAuth(), res = null;
+        if (authData !== null) {
```

Comment:
I realize you just copied this over from my existing code, but this can be simplified to:

```
if (rejectIfAuthDataIsNull && authData === null) {
  res = utils.reject("AUTH_REQUIRED");
}
else {
  res = utils.resolve(authData);
}
```

I think that reads a bit more easily.